File: Python/(Easy)AddBinary.py
# Add Binary

# Given two binary strings a and b, return their sum as a binary string.

# Example 1:
# Input: a = "11", b = "1"
# Output: "100"

# Example 2:
# Input: a = "1010", b = "1011"
# Output: "10101"


# Solution from Copilot (correct)
# Big O time complexity: O(log n)
class Solution(object):
    def addBinary(self, a, b):
        """
        :type a: str
        :type b: str
        :rtype: str
        """
        a = int(a, 2)
        b = int(b, 2)
        return bin(a+b)[2:]
    


# Solution from ChatGPT (correct)
# Big O time complexity: O(n)
class Solution(object):
    def addBinary(self, a, b):
        """
        :type a: str
        :type b: str
        :rtype: str
        """
        binary_sum = ""
        carry = 0
        
        i, j = len(a)-1, len(b)-1
        while i >= 0 or j >= 0:
            bit_a = int(a[i]) if i >= 0 else 0
            bit_b = int(b[j]) if j >= 0 else 0
            
            current_sum = bit_a + bit_b + carry
            
            binary_sum = str(current_sum % 2) + binary_sum
            carry = current_sum // 2
            
            i -= 1
            j -= 1
        
        if carry > 0:
            binary_sum = str(carry) + binary_sum
        
        return binary_sum

File: JS/(Easy)AddBinary.js
// Add Binary

// Given two binary strings a and b, return their sum as a binary string.

// Example 1:
// Input: a = "11", b = "1"
// Output: "100"

// Example 2:
// Input: a = "1010", b = "1011"
// Output: "10101"

/**
 * @param {string} a
 * @param {string} b
 * @return {string}
 */


// Solution from Copilot (correct)
// Big O time complexity: O(n)
var addBinary = function(a, b) {
    let result = "";
    let carry = 0;
    let i = a.length - 1;
    let j = b.length - 1;
    while (i >= 0 || j >= 0) {
        let sum = carry;
        if (i >= 0) {
            sum += parseInt(a[i]);
            i--;
        }
        if (j >= 0) {
            sum += parseInt(b[j]);
            j--;
        }
        if (sum % 2 === 0) {
            result = "0" + result;
        } else {
            result = "1" + result;
        }
        carry = sum > 1 ? 1 : 0;
    }
    if (carry === 1) {
        result = "1" + result;
    }
    return result;
};



// Solution from ChatGPT (correct)
// Big O time complexity: O(n)
var addBinary = function(a, b) {
    let carry = 0;
    let result = "";
    let i = a.length - 1;
    let j = b.length - 1;

    while (i >= 0 || j >= 0 || carry > 0) {
        let sum = carry;
        if (i >= 0) {
            sum += parseInt(a.charAt(i));
            i--;
        }
        if (j >= 0) {
            sum += parseInt(b.charAt(j));
            j--;
        }
        result = (sum % 2) + result;
        carry = Math.floor(sum / 2);
    }

    return result;
};